org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/generators/xml2relaxng/XML2RelaxNGGenerator.java

```diff
@@ -0,0 +1,269 @@
+/*******************************************************************************
+* Copyright (c) 2022 Red Hat Inc. and others.
+* All rights reserved. This program and the accompanying materials
+* which accompanies this distribution, and is available at
+* http://www.eclipse.org/legal/epl-v20.html
+*
+* SPDX-License-Identifier: EPL-2.0
+*
+* Contributors:
+*     Red Hat Inc. - initial API and implementation
+*******************************************************************************/
+package org.eclipse.lemminx.extensions.generators.xml2relaxng;
+
+import java.util.Collection;
+import java.util.Set;
+
+import org.eclipse.lemminx.dom.DOMAttr;
+import org.eclipse.lemminx.extensions.generators.AbstractXML2GrammarGenerator;
+import org.eclipse.lemminx.extensions.generators.AttributeDeclaration;
+import org.eclipse.lemminx.extensions.generators.AttributeDeclaration.DataType;
+import org.eclipse.lemminx.extensions.generators.Cardinality;
+import org.eclipse.lemminx.extensions.generators.ElementDeclaration;
+import org.eclipse.lemminx.extensions.generators.Grammar;
+import org.eclipse.lemminx.utils.StringUtils;
+import org.eclipse.lemminx.utils.XMLBuilder;
+import org.eclipse.lsp4j.jsonrpc.CancelChecker;
+
+/**
+ * File Generator implementation to generate RelaxNG from a given XML source.
+ *
+ */
+public class XML2RelaxNGGenerator extends AbstractXML2GrammarGenerator<RelaxNGGeneratorSettings> {
+
+	private static final String ELEMENT_ELT = "element"; //$NON-NLS-1$
+
+	private static final String NAME_ATTR = "name"; //$NON-NLS-1$
+
+	private static final String ONE_OR_MORE_ELT = "oneOrMore"; //$NON-NLS-1$
+
+	private static final String OPTIONAL_ELT = "optional"; //$NON-NLS-1$
+
+	private static final String TEXT_ELT = "text"; //$NON-NLS-1$
+
+	private static final String CHOICE_ELT = "choice";
+
+	private static final String ATTRIBUTE_ELT = "attribute";
+
+	private static final String VALUE_ELT = "value";
+
+	private static final String DATA_ELT = "data";
+
+	private static final String TYPE_ATTR = "type";
+
+	private static final String EMPTY_ELT = "empty";
+
+	private static final String MIXED_ELT = "mixed";
+
+	private static final String NAMESPACE_ATTR = "rng:ns";
+
+	private static final String DATATYPE_LIBRARY_ATTR = "rng:datatypeLibrary";
+
+	private static final String XML_SCHEMA_DATATYPES = "http://www.w3.org/2001/XMLSchema-datatypes";
+
+	private static final String INTERLEAVE_ELT = "interleave";
+
+	private static final String RELAXNG_NAMESPACE_URI = "http://relaxng.org/ns/structure/1.0"; //$NON-NLS-1$
+
+	@Override
+	protected void generate(Grammar grammar, RelaxNGGeneratorSettings settings, XMLBuilder schema,
+			CancelChecker cancelChecker) {
+		String prefix = "rng";
+		// <?xml version="1.0" encoding="UTF-8"?>
+		schema.startPrologOrPI("xml");
+		schema.addContent(" version=\"1.0\" encoding=\"UTF-8\"");
+		schema.endPrologOrPI();
+		// <element name="rootElementName" xmlns:rng="http://relaxng.org/ns/structure/1.0">
```

Comment:
Please generate the same content than snippet

* without rng prefix
* with grammar element
* with define pattern